### PR TITLE
Add PNG export for Notebook's visualization blocks

### PR DESCRIPTION
**Pull request by vtfg**
Adds #18

Currently, exporting works for all Briefer's visualization blocks and Python's graph outputs in the Notebook tab.

### TODO
- [x] Show button when Visualization Controls are not hidden
- [x] Show button in Python's graph outputs
- [x] ~~Identify when the `renderer` is `svg` and code a solution for downloading it~~
  - The only case where the `renderer` is `svg` is in the Dashboard's block list, where the export button is not visible
- [x] Hide button if chart doesn't support export
- [x] ~~Support Internet Explorer and Opera Mini~~
  - [Can I Use `a[download]`](https://caniuse.com/?search=a.download)
  - Briefer itself doesn't support Internet Explorer, so we don't need to do this
- [x] Simplify the export function (maybe by downloading a library or moving the code to a helper file)

### CODE REVIEW

**Comment by lucasfcosta** [comment]
Thanks for this! Btw, I saw you marked this as draft - lmk if you need some input from my side before finalizing it. Happy to help in any way I can.

**Comment by vtfg** [comment]
I'm studying the code base in order to add this export button to Python's graph outputs.

As I mentioned, now the "PNG" button is only visible when the visualization controls are hidden due to the graph shrinking (responsiveness).

The only two ways that I imagine to "fix" this (and making the button visible in both states) is either by closing, exporting and opening the sidebar again in the blink of an eye, or by copying the whole graph's canvas to a hidden div with normal size (that'll be deleted when the export is done).

Do you have a better solution in mind?

File: apps/web/src/utils/file.ts [hunk 1]
```diff
@@ -19,3 +19,20 @@ export function readFile(
...
 }
+
+export function downloadFile(url: string, name: string) {
+  const downloadLink = document.createElement('a')
+
+  // some browsers don't support the a[download] solution
+  if (typeof downloadLink.download === 'undefined') {
+    // 'window.location.href = downloadLink' no long works due to some Chrome policies update
+    return
+  }
```
Comment (lucasfcosta) [inline apps/web/src/utils/file.ts:30]: I think that we could just always use the append and remove approach if that works on all browsers. That way we can remove this `if` statement here - I'm not fussy about using this new feature in particular.

Thanks for going after this though, I learned something new! I always just used the append approach.

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx
Comment (lucasfcosta) [inline apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:1]: I'm glad you also included the `PythonOutput` in this PR - I hadn't mentioned it in the issue but you were quite attentive here! Excellent work! 🙌

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx [hunk 2]
```diff
@@ -106,16 +107,33 @@ interface ItemProps {
   canFixWithAI: boolean
 }
 export function PythonOutput(props: ItemProps) {
+  const onExportToPNG = () => {
+    if (props.output.type !== 'image' || props.output.format !== 'png') return
+
+    downloadFile(
+      `data:image/${props.output.format};base64, ${props.output.data}`,
+      'Generated Python image'
```
Comment (lucasfcosta) [inline apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:115]: Isn't this name supposed to be the filename? I think we could use the block's ID followed by `.png` here

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx [hunk 3]
```diff
@@ -106,16 +107,33 @@ interface ItemProps {
...
-              src={`data:image/${props.output.format};base64, ${props.output.data}`}
-            />
+            <>
+              <img
+                className="printable-block"
+                alt="generated image"
+                src={`data:image/${props.output.format};base64, ${props.output.data}`}
+              />
+              <button
+                className="absolute bottom-0 bg-white rounded-md border border-gray-200 p-1 hover:bg-gray-50 z-10 right-0 text-xs text-gray-400"
```
Comment (lucasfcosta) [inline apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:131]: In my local tests this button's borders look slightly weird because it's covering the block's border (and all sides are rounded).

```suggestion
                className="absolute bottom-[1px] right-[1px] bottom-0 bg-white rounded-tl-md rounded-br-md border border-gray-200 p-1 hover:bg-gray-50 z-10 right-0 text-xs text-gray-400"
```

This fixes it (I moved it a px in each axis + fixed the border radius)

Screenshot 2024-09-15 at 16 26 25

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx [hunk 4]
```diff
@@ -106,16 +107,33 @@ interface ItemProps {
   canFixWithAI: boolean
 }
 export function PythonOutput(props: ItemProps) {
+  const onExportToPNG = () => {
```
Comment (lucasfcosta) [inline apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:110]: There's a small issue here, which is that when you have more than one image output, clicking the `PNG` button only downloads the last one.

To reproduce, run this code and try downloading the PNG:

```python
import matplotlib.pyplot as plt
import numpy as np

# Sample data
x = np.linspace(0, 10, 100)
y = np.sin(x)
y2 = np.cos(x)

# Plotting sin(x) as the first output
plt.figure(figsize=(10, 6))
plt.plot(x, y, label='sin(x)')
plt.title('Plot of sin(x)')
plt.xlabel('x')
plt.ylabel('sin(x)')
plt.legend()
plt.grid(True)
plt.show()

# Plotting cos(x) as the second output
plt.figure(figsize=(10, 6))
plt.plot(x, y2, label='cos(x)', color='r')
plt.title('Plot of cos(x)')
plt.xlabel('x')
plt.ylabel('cos(x)')
plt.legend()
plt.grid(True)
plt.show()
```

By the way, this is not a big deal, I'm okay with merging with this issue and opening a separate issue to handle multi-output exports. I'll leave it up to you.

File: apps/web/src/components/v2Editor/customBlocks/visualization/index.tsx [hunk 5]
```diff
@@ -202,6 +203,34 @@ function VisualizationBlock(props: Props) {
...
+  const onExportToPNG = async () => {
+    // we don't need to check if props.renderer is undefined because the application sets as 'canvas' in this case
+    if (
+      props.renderer === 'svg' ||
+      chartType === 'number' ||
+      chartType === 'trend'
+    )
+      return
+
+    // if the controls are visible the canvas got shrinked, making the export smaller
```
Comment (lucasfcosta) [inline apps/web/src/components/v2Editor/customBlocks/visualization/index.tsx:215]: This is fine! I'm okay with that! Great idea here.

Minor grammatical nit:

```suggestion
    // if the controls are visible the canvas shrinks, making the export smaller
```

**Review by lucasfcosta — COMMENTED** [review]
Thanks so much for this PR @vtfg! It's looking awesome.

Great work also considering the `PythonOutput` block 🙌 

I just added a preliminary review to help you out - I'm okay with what you mentioned about hiding the controls when clicking the button btw.

**Comment by lucasfcosta** [comment]
Btw just one more comment: the PythonOutput still shows the `PNG` button while in the dashboard. IMO we should hide this button in the dashboard view.

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx (hunk 4 above)
Comment (vtfg) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:110]: Oh, didn't knew we could do that!

I'll try to handle multiple outputs in this PR.

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx (hunk 3 above)
Comment (vtfg) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:131]: I made the button float because of this weird Windows' scrollbar. And now I think it makes more sense due to the multi-output thing.

Here is a preview of it:

image

File: apps/web/src/utils/file.ts (hunk 1 above)
Comment (vtfg) [reply apps/web/src/utils/file.ts]: I tested and Briefer doesn't work on Internet Explorer. `if` statement removed.

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx (hunk 2 above)
Comment (vtfg) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx]: Done.

Adding, we don't need to add `.png` in the file name due to the MIME type (`props.output.format`) defined in the file url.

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx (hunk 4 above)
Comment (vtfg) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:114]: Fixed!

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx (hunk 3 above)
Comment (lucasfcosta) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx]: @vtfg thanks for this! The issue here is that it will look weird on Mac OS, which doesn't show the scrollbar (and most of our users are on Macs too)

 I think that the 1px float and the rounded borders are fine and what we should actually do is handle the scrollbars on Windows separately. In fact, they should not appear on Windows too (we have done this elsewhere).
 
 What do you think of adding these CSS changes and then I'd love it if you want to tackle updating the scrollbars so they don't appear on Windows (I use a mac, so I have to go through a lot of work to test on Windows, if you have one that'd be easier).

Comment (vtfg) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx]: @lucasfcosta I've commited the suggested CSS changes. 

I don't know if they'll have the expected impact because of the padding applied to the collapsable component, making these buttons not too close to the container's border (at least on Windows).

image

**Comment by lucasfcosta** [comment]
Thank you so much @vtfg! I see what you mean now. I didn't understand this was moved to a different container within the Python output. I'll merge as is and just do a very minor change on top.

Thank you so much for working on this. It was fantastic, and thanks so much for taking our suggestions into account.

Do you wanna try going for https://github.com/briefercloud/briefer/issues/72 next now that you have context? Happy to help you out with the next contribution!

File: apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx [hunk 6]
```diff
@@ -80,6 +82,7 @@ export function PythonOutputs(props: Props) {
           key={i}
           className={clsx(
             ['plotly'].includes(output.type) ? 'flex-grow' : '',
+            !props.isDashboardView ? 'flex flex-col items-end' : '',
```
Comment (vieiralucas) [inline apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:85]: Why is this needed? This makes all output align to the right, including text, it's looking weird now.
image

Comment (vtfg) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:85]: Oh, that's bad. It was made to move the PNG button (with relative position) to the right.

I can fix that right now in a new PR. Or if you want to, move these classes (with the ternary) to a new div inside the 'png' case in the PythonOutput component.

Comment (vtfg) [reply apps/web/src/components/v2Editor/customBlocks/python/PythonOutput.tsx:85]: @vieiralucas Fixed in #82.
